File: app/src/main/java/com/baijum/ukufretboard/viewmodel/ChordLibraryViewModel.kt
package com.baijum.ukufretboard.viewmodel

import androidx.lifecycle.ViewModel
import com.baijum.ukufretboard.data.ChordCategory
import com.baijum.ukufretboard.data.ChordFormula
import com.baijum.ukufretboard.data.ChordFormulas
import com.baijum.ukufretboard.data.Notes
import com.baijum.ukufretboard.data.VoicingGenerator
import com.baijum.ukufretboard.domain.ChordVoicing
import com.baijum.ukufretboard.domain.Transpose
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update

/**
 * Immutable UI state for the chord library tab.
 *
 * @property selectedRoot The pitch class (0–11) of the currently selected root note.
 * @property selectedCategory The currently selected chord category filter.
 * @property selectedFormula The currently selected chord formula, or null if none is chosen.
 * @property voicings The generated voicings for the current root + formula selection.
 */
data class ChordLibraryUiState(
    val selectedRoot: Int = 0,
    val selectedCategory: ChordCategory = ChordCategory.TRIAD,
    val selectedFormula: ChordFormula? = null,
    val voicings: List<ChordVoicing> = emptyList(),
)

/**
 * ViewModel for the chord library tab.
 *
 * Manages the selection of root note, chord category, and chord formula,
 * and generates playable voicings via [VoicingGenerator] whenever the
 * selection changes.
 */
class ChordLibraryViewModel : ViewModel() {

    /** The tuning used for voicing generation. */
    private var tuning = FretboardViewModel.STANDARD_TUNING

    /** Whether the voicing generator may include voicings with muted strings. */
    private var allowMutedStrings: Boolean = false

    private val _uiState = MutableStateFlow(
        ChordLibraryUiState().let { initial ->
            // Auto-select the first formula in the default category
            val firstFormula = ChordFormulas.BY_CATEGORY[initial.selectedCategory]?.firstOrNull()
            if (firstFormula != null) {
                initial.copy(
                    selectedFormula = firstFormula,
                    voicings = VoicingGenerator.generate(
                        rootPitchClass = initial.selectedRoot,
                        formula = firstFormula,
                        tuning = tuning,
                    ),
                )
            } else {
                initial
            }
        }
    )

    /** Observable UI state for the chord library tab. */
    val uiState: StateFlow<ChordLibraryUiState> = _uiState.asStateFlow()

    /**
     * Updates the tuning used by voicing generation and refreshes current results.
     */
    fun setTuning(tuning: List<UkuleleString>) {
        if (this.tuning == tuning) return
        this.tuning = tuning
        _uiState.update { current ->
            current.copy(
                voicings = generateVoicings(current.selectedRoot, current.selectedFormula),
            )
        }
    }

    /**
     * Selects a root note and regenerates voicings.
     *
     * @param pitchClass The pitch class (0–11) of the root note.
     */
    fun selectRoot(pitchClass: Int) {
        _uiState.update { current ->
            current.copy(
                selectedRoot = pitchClass,
                voicings = generateVoicings(pitchClass, current.selectedFormula),
            )
        }
    }

    /**
     * Selects a chord category and auto-selects its first formula.
     *
     * @param category The [ChordCategory] to switch to.
     */
    fun selectCategory(category: ChordCategory) {
        val firstFormula = ChordFormulas.BY_CATEGORY[category]?.firstOrNull()
        _uiState.update { current ->
            current.copy(
                selectedCategory = category,
                selectedFormula = firstFormula,
                voicings = generateVoicings(current.selectedRoot, firstFormula),
            )
        }
    }

    /**
     * Selects a specific chord formula and regenerates voicings.
     *
     * @param formula The [ChordFormula] to select.
     */
    fun selectFormula(formula: ChordFormula) {
        _uiState.update { current ->
            current.copy(
                selectedFormula = formula,
                voicings = generateVoicings(current.selectedRoot, formula),
            )
        }
    }

    /**
     * Transposes the current root by [semitones] (+1 up, -1 down) and
     * regenerates voicings.
     */
    fun transpose(semitones: Int) {
        _uiState.update { current ->
            val newRoot = Transpose.transposePitchClass(current.selectedRoot, semitones)
            current.copy(
                selectedRoot = newRoot,
                voicings = generateVoicings(newRoot, current.selectedFormula),
            )
        }
    }

    /**
     * Returns the display name for the currently selected chord
     * (e.g., "Cm7", "G", "F#dim").
     */
    fun currentChordName(): String {
        val state = _uiState.value
        val rootName = Notes.pitchClassToName(state.selectedRoot)
        val symbol = state.selectedFormula?.symbol ?: ""
        return "$rootName$symbol"
    }

    /**
     * Updates the muted-string setting and regenerates voicings if the value changed.
     *
     * Called from the UI layer when the "Allow Muted Strings" setting changes.
     */
    fun setAllowMutedStrings(allowed: Boolean) {
        if (allowMutedStrings == allowed) return
        allowMutedStrings = allowed
        _uiState.update { current ->
            current.copy(
                voicings = generateVoicings(current.selectedRoot, current.selectedFormula),
            )
        }
    }

    private fun generateVoicings(rootPitchClass: Int, formula: ChordFormula?): List<ChordVoicing> {
        if (formula == null) return emptyList()
        return VoicingGenerator.generate(
            rootPitchClass = rootPitchClass,
            formula = formula,
            tuning = tuning,
            allowMutedStrings = allowMutedStrings,
        )
    }
}

File: iosApp/UkuleleCompanion/ViewModels/ChordLibraryViewModel.swift
import Foundation
import shared

/// View model for the chord library browser.
@MainActor
final class ChordLibraryViewModel: ObservableObject {
    @Published var selectedRoot: Int32 = 0
    @Published var selectedCategory: ChordCategory = .triad
    @Published var selectedFormula: ChordFormula?
    @Published var voicings: [ChordVoicing] = []
    @Published var searchQuery: String = ""
    @Published var searchResults: [ChordSearchResult] = []

    private let tuning: [shared.UkuleleString] = {
        let t = UkuleleTuning.highG
        return (0..<4).map { i in
            shared.UkuleleString(
                name: t.stringNames[i] as! String,
                openPitchClass: (t.pitchClasses[i] as! NSNumber).int32Value,
                octave: (t.octaves[i] as! NSNumber).int32Value
            )
        }
    }()

    init() {
        let formulas = formulasForCategory(.triad)
        if let first = formulas.first {
            selectedFormula = first
            voicings = generateVoicings(root: 0, formula: first)
        }
    }

    var rootNoteNames: [String] {
        (0..<Int32(12)).map { Notes.shared.pitchClassToName(pitchClass: $0) }
    }

    func formulasForCategory(_ category: ChordCategory) -> [ChordFormula] {
        guard let list = ChordFormulas.shared.BY_CATEGORY[category] else { return [] }
        return list as! [ChordFormula]
    }

    var currentChordName: String {
        let rootName = Notes.shared.pitchClassToName(pitchClass: selectedRoot)
        let symbol = selectedFormula?.symbol ?? ""
        return rootName + symbol
    }

    func selectRoot(_ pitchClass: Int32) {
        selectedRoot = pitchClass
        regenerateVoicings()
    }

    func selectCategory(_ category: ChordCategory) {
        selectedCategory = category
        let formulas = formulasForCategory(category)
        selectedFormula = formulas.first
        regenerateVoicings()
    }

    func selectFormula(_ formula: ChordFormula) {
        selectedFormula = formula
        regenerateVoicings()
    }

    // MARK: - Chord Search

    func updateSearchQuery(_ query: String) {
        searchQuery = query
        searchResults = ChordNameParser.shared.suggestions(query: query) as [ChordSearchResult]
    }

    /// Navigates to the chord described by the search result.
    /// Returns the target inversion if the result is a slash chord, nil otherwise.
    func selectSearchResult(_ result: ChordSearchResult) -> ChordInfo.Inversion? {
        selectRoot(result.rootPitchClass)
        selectCategory(result.formula.category)
        selectFormula(result.formula)
        searchQuery = ""
        searchResults = []
        return result.inversion
    }

    private func regenerateVoicings() {
        guard let formula = selectedFormula else {
            voicings = []
            return
        }
        voicings = generateVoicings(root: selectedRoot, formula: formula)
    }

    private func generateVoicings(root: Int32, formula: ChordFormula) -> [ChordVoicing] {
        let result = VoicingGenerator.shared.generate(
            rootPitchClass: root,
            formula: formula,
            tuning: tuning,
            allowMutedStrings: false
        )
        return result as! [ChordVoicing]
    }
}
